File: cpp/organism/Organism.cpp
//
// Created by Filip on 12/04/2025.
//

#include "Organism.h"
#include "../world/World.h"
#include <algorithm>

Organism::Organism(Point point, std::string symbol, int strength,
                   int initiative, World &world)
    : position(point), symbol(symbol), strength(strength),
      initiative(initiative), age(0), world(world) {};

Organism::~Organism() {
    // Pusta definicja destruktora
}

Point Organism::getPosition() { return position; }

Point Organism::setPosition(Point newPosition) {
    position = newPosition;

    return position;
}

int Organism::getStrength() const { return strength; }

void Organism::setStrength(int strength) { this->strength = strength; }
void Organism::setInitiative(int initiative) { this->initiative = initiative; }

int Organism::getInitiative() const { return initiative; }

int Organism::getAge() { return age; }

int Organism::setAge(int newAge) {
    age = newAge;
    return age;
}

int Organism::increaseAge() {
    age++;
    return age;
}

std::string Organism::getSymbol() const { return symbol; }

Point Organism::getNewPosition() {
    std::vector<Point> directions = {Point(0, 1), Point(1, 0), Point(0, -1),
                                     Point(-1, 0), Point(1, 1), Point(1, -1),
                                     Point(-1, 1), Point(-1, -1)};
    Point newPosition;

    int maxAttempts = directions.size();// Maksymalna liczba prób
    int attempts = 0;

    while (!directions.empty() && attempts < maxAttempts) {

        std::uniform_int_distribution<> dist(0, directions.size() - 1);
        int index = dist(world.getRandomEngine());
        newPosition = position + directions[index];
        directions.erase(directions.begin() + index);// Usuń sprawdzony kierunek
        attempts++;

        if (world.isInBounds(newPosition)) {
            return newPosition;
        }
    }

    // Jeśli nie znaleziono wolnej przestrzeni, zwróć aktualną pozycję
    return position;
}

void Organism::action() {

    increaseAge();
    Point newPosition = getNewPosition();
    move(newPosition);
}

void Organism::move(Point newPosition) {
    bool isThisOrganismDeleted;
    // Sprawdź, czy nowa pozycja jest zajęta
    bool positionOccupied = false;
    for (auto &other: world.getOrganisms()) {
        if (other != nullptr && other.get() != this &&
            other->getPosition() == newPosition) {
            positionOccupied = true;

            isThisOrganismDeleted = other->collision(*this);
            if (!isThisOrganismDeleted) {
                break;
            }
        }
    }

    // Jeśli pozycja nie jest zajęta, wykonaj ruch
    if (!isThisOrganismDeleted && !world.isOccupied(newPosition)) {
        position = newPosition;
    }
}
void Organism::die() {
    // Pusta definicja metody
    world.removeOrganism(this);
}

Point Organism::getPreviousPosition() const { return previousPosition; }

void Organism::changeSymbol(const std::string &newSymbol) {
    symbol = newSymbol;
}

File: java/src/organism/Organism.java
package organism;

import logger.EventLogger;
import world.World;

import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class Organism {
    protected int x, y;
    protected String symbol;
    protected int strength;
    protected int initiative;
    protected int age;
    protected int previousX, previousY;
    protected World world;

    public Organism(int x, int y, String symbol, int strength, int initiative, World world) {
        this.x = x;
        this.y = y;
        this.symbol = symbol;
        this.strength = strength;
        this.initiative = initiative;
        this.age = 0;
        this.world = world;
        this.previousX = x;
        this.previousY = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void setPosition(int x, int y) {
        this.previousX = this.x;
        this.previousY = this.y;
        this.x = x;
        this.y = y;
    }

    public int getStrength() {
        return strength;
    }

    public void setStrength(int strength) {
        this.strength = strength;
    }

    public int getInitiative() {
        return initiative;
    }

    public void setInitiative(int initiative) {
        this.initiative = initiative;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public int increaseAge() {
        return ++age;
    }

    public String getSymbol() {
        return symbol;
    }

    public void changeSymbol(String newSymbol) {
        this.symbol = newSymbol;
    }

    public int getPreviousX() {
        return previousX;
    }

    public int getPreviousY() {
        return previousY;
    }

    // Returns a random valid neighbor position (4 or 8 directions)
    protected int[] getNewPosition() {
        List<int[]> directions = new ArrayList<>();
        directions.add(new int[]{0, 1});    // Up
        directions.add(new int[]{1, 0});    // Right
        directions.add(new int[]{0, -1});   // Down
        directions.add(new int[]{-1, 0});   // Left
        directions.add(new int[]{1, 1});    // Up-Right
        directions.add(new int[]{1, -1});   // Down-Right
        directions.add(new int[]{-1, 1});   // Up-Left
        directions.add(new int[]{-1, -1});  // Down-Left

        Random rand = new Random();
        while (!directions.isEmpty()) {
            int idx = rand.nextInt(directions.size());
            int[] dir = directions.remove(idx);
            int nx = x + dir[0];
            int ny = y + dir[1];
            if (world.isInside(nx, ny)) {
                return new int[]{nx, ny};
            }
        }
        return new int[]{x, y};
    }

    public void action() {
            int[] newPos = getNewPosition();
            move(newPos[0], newPos[1]);

        increaseAge();
    }

    public void move(int newX, int newY) {
        EventLogger.getInstance().log("Moved " + symbol + " form (" + x + ", " + y + ") to (" + newX + ", " + newY + ")");
        Organism other = world.getOrganismAt(newX, newY);
        boolean isThisOrganismDeleted = false;
        if (other != null && other != this) {
            EventLogger.getInstance().log("Collision between " + getClass().getName() + " and " + other.getClass().getName() + " at (" + newX + ", " + newY + ")");
            isThisOrganismDeleted = other.collision(this);
        }
        if (!isThisOrganismDeleted && world.getOrganismAt(newX, newY) == null) {
            setPosition(newX, newY);
        }
    }

    public void die() {
        EventLogger.getInstance().log("Organism " + symbol + " died at (" + x + ", " + y + ")");
        world.removeOrganism(this);
    }

    // Abstract: must be implemented in subclasses
    public abstract boolean collision(Organism attacker);
}
